### PR TITLE
feat: Break out `custodial-ownership-token` directory in maple-core to be a submodule dapptools project (SC-3052)

### PR DESCRIPTION
**Pull request by satyamakgec**

### CODE REVIEW

**Comment by lucas-manuel** [comment]
Tests not compiling @satyamakgec

File: config/ci.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+  "language": "Solidity",
+  "sources": {
+    "contracts/interfaces/IERC2258.sol": {
+      "urls": ["contracts/interfaces/IERC2258.sol"]
+    },
```
Comment (lucas-manuel) [inline config/ci.json]: Interface not needed

File: config/dev.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+  "language": "Solidity",
+  "sources": {
+     "contracts/interfaces/IERC2258.sol": {
+      "urls": ["contracts/interfaces/IERC2258.sol"]
+    },
```
Comment (lucas-manuel) [inline config/dev.json]: Interface not needed

File: config/prod.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
+  "language": "Solidity",
+  "sources": {
+    "contracts/interfaces/IERC2258.sol": {
+      "urls": ["contracts/interfaces/IERC2258.sol"]
+    },
```
Comment (lucas-manuel) [inline config/prod.json]: Interface not needed

File: contracts/test/ERC2258.t.sol [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import { ERC2258Account } from "./accounts/ERC2258Account.sol";
+
+import { IERC20 } from "../../modules/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+
+interface Hevm {
+
+    function warp(uint256) external;
+    function store(address,bytes32,bytes32) external;
+
+}
```
Comment (lucas-manuel) [inline contracts/test/ERC2258.t.sol]:
```
interface Hevm {

    function warp(uint256) external;

    function store(address,bytes32,bytes32) external;

}
```

File: contracts/test/ERC2258.t.sol [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+contract ERC2258Test is DSTest {
+
+    ERC2258     erc2258;
+    Custodian       cus;
+    Custodian      cus2;
+    ERC2258Account  act;
+    ERC2258Account act2;
+    Hevm           hevm;
+
```
Comment (lucas-manuel) [inline contracts/test/ERC2258.t.sol]: Extra line

File: Makefile [hunk 6]
```diff
@@ -1,4 +1,5 @@
-all    :; dapp build
-clean  :; dapp clean
-test   :; dapp test
-deploy :; dapp create CustodialOwnershipToken
+prod    :; ./build.sh -c ./config/prod.json
+dev     :; ./build.sh -c ./config/dev.json
+ci      :; ./build.sh -c ./config/ci.json
+clean   :; dapp clean
+test    :; ./test.sh
```
Comment (deluca-mike) [inline Makefile]: new line needed at the end of the file

File: config/ci.json (hunk 1 above)
Comment (deluca-mike) [inline config/ci.json]: Can be removed at its not a unique entrypoint with respect to ci.

File: config/dev.json (hunk 2 above)
Comment (deluca-mike) [inline config/dev.json]: Can be removed at its not a unique entrypoint with respect to dev.

File: config/prod.json (hunk 3 above)
Comment (deluca-mike) [inline config/prod.json]: Can be removed at its not a unique entrypoint with respect to prod.

File: contracts/test/ERC2258.t.sol [hunk 7]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity >=0.6.11;
+pragma experimental ABIEncoderV2;
+
+import { ERC2258 } from "../ERC2258.sol";
+import { DSTest } from "../../modules/ds-test/src/test.sol";
+import { Custodian } from "./accounts/Custodian.sol";
+import { ERC2258Account } from "./accounts/ERC2258Account.sol";
+
+import { IERC20 } from "../../modules/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]:
```suggestion
import { DSTest } from "../../modules/ds-test/src/test.sol";
import { IERC20 } from "../../modules/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";

import { ERC2258 } from "../ERC2258.sol";

import { Custodian }      from "./accounts/Custodian.sol";
import { ERC2258Account } from "./accounts/ERC2258Account.sol";
```

File: contracts/test/ERC2258.t.sol [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+        hevm.store(
+            addr,
+            keccak256(abi.encode(account, slot)), // Mint tokens
+            bytes32(bal + amt)
+        );
+
+        assertEq(IERC20(addr).balanceOf(account), bal + amt, "Balance slot is wrong");  // Assert new balance
+    }
+}
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]: new line needed at the end of the file.

File: contracts/test/ERC2258.t.sol [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    Custodian       cus;
+    Custodian      cus2;
+    ERC2258Account  act;
+    ERC2258Account act2;
+    Hevm           hevm;
+    
+
+    uint256 constant WAD = 10 ** 18;
+
+    constructor() public { hevm = Hevm(address(bytes20(uint160(uint256(keccak256("hevm cheat code")))))); }
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]:
```suggestion
    constructor() public {
        hevm = Hevm(address(bytes20(uint160(uint256(keccak256("hevm cheat code"))))));
    }
```

File: contracts/test/ERC2258.t.sol [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    function setUp() public {
+        erc2258 = new ERC2258("Token", "TKN");
+        cus     = new Custodian();
+        cus2    = new Custodian();
+        act     = new ERC2258Account();
+        act2    = new ERC2258Account();
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]:
```suggestion
        token = new ERC2258("Token", "TKN");
        cus1  = new Custodian();
        cus2  = new Custodian();
        act1  = new ERC2258Account();
        act2  = new ERC2258Account();
```

File: contracts/test/ERC2258.t.sol [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+    ERC2258     erc2258;
+    Custodian       cus;
+    Custodian      cus2;
+    ERC2258Account  act;
+    ERC2258Account act2;
+    Hevm           hevm;
+    
+
+    uint256 constant WAD = 10 ** 18;
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]: No need for this as `WAD` is an irrelevant concept with respect to ERC2258.

File: contracts/test/ERC2258.t.sol [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        uint256 slot = 0;
+        uint256 bal  = IERC20(addr).balanceOf(account);
+
+        hevm.store(
+            addr,
+            keccak256(abi.encode(account, slot)), // Mint tokens
+            bytes32(bal + amt)
+        );
+
+        assertEq(IERC20(addr).balanceOf(account), bal + amt, "Balance slot is wrong");  // Assert new balance
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]: `mint` isn't a test and minting isn't functionality being tested, so this assertion is out of scope.

File: contracts/test/ERC2258.t.sol [hunk 13]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity >=0.6.11;
+pragma experimental ABIEncoderV2;
```
Comment (deluca-mike) [inline contracts/test/ERC2258.t.sol]: Not needed.

File: contracts/test/accounts/Custodian.sol [hunk 14]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity >=0.6.11;
+pragma experimental ABIEncoderV2;
```
Comment (deluca-mike) [inline contracts/test/accounts/Custodian.sol]: Not needed.

File: contracts/test/accounts/ERC2258Account.sol [hunk 15]
```diff
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: AGPL-3.0-or-later
 pragma solidity >=0.6.11;
+pragma experimental ABIEncoderV2;
```
Comment (deluca-mike) [inline contracts/test/accounts/ERC2258Account.sol]: Not needed.

**Review by deluca-mike — CHANGES_REQUESTED** [review]
`modules/openzeppelin-contracts` is not the correct `v3.3.0` commit

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: contracts/ERC2258.sol [hunk 16]
```diff
@@ -34,11 +34,21 @@ contract ERC2258 is IERC2258, ERC20 {
...
-        uint256 newTotalAllowance          = totalCustodyAllowance[from].sub(amount);
-        totalCustodyAllowance[from]        = newTotalAllowance;
-        
-        emit CustodyTransfer(msg.sender, from, to, amount);
+
         emit CustodyAllowanceChanged(from, msg.sender, oldAllowance, newAllowance);
+
+        if (from == to) {
+            totalCustodyAllowance[from] = totalCustodyAllowance[from].sub(amount);
+        } else {
```
Comment (lucas-manuel) [inline contracts/ERC2258.sol]: Undo this

**Review by deluca-mike — APPROVED** [review]

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**